src/sage/crypto/historical/enigma.py

```diff
@@ -0,0 +1,1148 @@
+            sage: d = Reflector([24, 17, 20, 7, 16, 18, 11, 3, 15, 23, 13, 6,
+            ....: 14, 10, 12, 8, 4, 1, 5, 25, 2, 22, 21, 9, 0, 19])
+            sage: e = Enigma([a, b, c], d, None)
+            sage: e.en_de_crypt("HELLO")
+            'GCIID'
+
+        The text can not contain any other character than the letters A - Z::
+
+            sage: e.en_de_crypt("HELLO HOW ARE YOU")
+            Traceback (most recent call last):
+            ...
+            ValueError: Invalid letter: Input can only contain letters A - Z
+
+        However, lowercase letters are automatically converted to uppercase
+        letters::
+
+            sage: e.en_de_crypt("hellohowareyou")
+            'XPVIDDZIXUIDVA'
+        """
+
+        result = ""
+        message = str(text).upper()
+
+        for letter in message:
+            predecessor_tuple: tuple[str, bool] = letter, True
+
+            if self._plugboard is not None:
+                predecessor_tuple = (
+                    self._plugboard._process_letter(predecessor_tuple[0]),
+                    True,
+                )
+
+            for rotor in self._rotors:
+                predecessor_tuple = rotor._process_letter_forward(
+                    predecessor_tuple[0], predecessor_tuple[1]
+                )
+
+            predecessor_tuple = (
+                self._reflector._process_letter(predecessor_tuple[0]),
+                False,
+            )
+
+            for rotor in reversed(self._rotors):
+                predecessor_tuple = (
+                    rotor._process_letter_backward(predecessor_tuple[0]),
+                    False,
+                )
+
+            if self._plugboard is not None:
+                predecessor_tuple = (
+                    self._plugboard._process_letter(predecessor_tuple[0], False),
+                    False,
+                )
+
+            result += predecessor_tuple[0]
+
+        return result
+
+    def duplicate(self):
+        r"""
+        Creates a copy of the current enigma. Note that it will not be a 1:1
+        copy of the enigma, since the initial positions of the new rotors will
+        be the current positions of the rotors in this enigma.
+
+        OUTPUT: Copy of this enigma
+
+        EXAMPLES::
+
+            sage: e = Enigma.create([5, 3, 2], [5, 24, 11], [2, 4, 24], 2)
+            sage: c = e.duplicate()
+            sage: text = "Hello".upper()
+            sage: encrypted_text = e.en_de_crypt(text)
+            sage: decrypted_text = c.en_de_crypt(encrypted_text)
+            sage: decrypted_text == text
+            True
+        """
+
+        # since rotors have a own state (position) we need a deep copy
+        new_rotors: list[Rotor] = []
+        for rotor in self._rotors:
+            new_rotors.append(rotor.duplicate())
+
+        # plugboard and reflector are static permutations anyway
+        return Enigma(new_rotors, self._reflector, self._plugboard)
+
+    def __repr__(self) -> str:
```

Comment:
Classes inheriting from `SageObject` should define `_repr_` instead of `__repr__`

Reply:
Classes should also define `_latex_`